devito/core/gpu_openacc.py

```diff
@@ -52,6 +86,78 @@ class DeviceOpenACCDataManager(DeviceOpenMPDataManager):
 
     _Parallelizer = DeviceAccizer
 
+    def _alloc_array_on_high_bw_mem(self, obj, storage):
+        """Allocate an Array in the high bandwidth memory."""
+        if obj in storage._high_bw_mem:
+            return
+
+        size_trunkated = "".join("[%s]" % i for i in obj.symbolic_shape[1:])
+        decl = c.Value(obj._C_typedata, "(*%s)%s" % (obj.name, size_trunkated))
+        cast = "(%s (*)%s)" % (obj._C_typedata, size_trunkated)
+        size_full = prod(obj.symbolic_shape)
+        alloc = "%s acc_malloc(sizeof(%s[%s]))" % (cast, obj._C_typedata, size_full)
+        init = c.Initializer(decl, alloc)
+
+        free = c.Statement('acc_free(%s)' % obj.name)
+
+        storage._high_bw_mem[obj] = (None, init, free)
+
+
+@iet_pass
+def initialize(iet, **kwargs):
+    """
+    Initialize the OpenACC environment.
+    """
+
+    @singledispatch
```

Comment:
Why is this needed?

Reply:
all IET passes are applied, individually, to all Callables in the IET. singledispatch is the software engineering pattern that allows you to specialize a pass for a specific type of Callable. In this case, it is used to make the pass such that with all non-Operator Callables it is essentially a no-op

`_initialize` is for initialization of the OpenACC runtime and it should be performed only in the Kernel function

Reply:
let me elaborate:

the reason I use the `singledispatch` is to make sure that that specific IET pass has effect only on the main function, while leaving all other Callables untouched

```
int Kernel(...) {
 ...
  haloupdate(...)
  bf0(...)
}
haloupdate(...) {
  ...
}
bf0(...) {
  ...
}
```

I want the pass to be applied only to
```
int Kernel(...) {
 ...
}
```